File: integration/node/minima-client.js
/**
 * Minima RPC Client for Node.js
 *
 * Safe, typed interface to a Minima node's RPC endpoint.
 * Normalizes responses so the default API path prevents common mistakes.
 *
 * @example
 * import { MinimaClient } from './minima-client.js';
 *
 * const client = new MinimaClient();
 * const bal = await client.balance();       // normalized, no 'total' trap
 * const status = await client.status();
 * const tx = await client.send('MxG08...', 1);
 */

class MinimaError extends Error {
  constructor(message) {
    super(message);
    this.name = 'MinimaError';
  }
}

class MinimaConnectionError extends Error {
  constructor(message) {
    super(message);
    this.name = 'MinimaConnectionError';
  }
}

class MinimaClient {
  /**
   * @param {Object} [options]
   * @param {string} [options.host='localhost']
   * @param {number} [options.port=9005]
   * @param {number} [options.retries=3]
   * @param {number} [options.retryDelay=1000] - Base retry delay in ms
   * @param {number} [options.timeout=30000] - Request timeout in ms
   */
  constructor({ host = 'localhost', port = 9005, retries = 3, retryDelay = 1000, timeout = 30000 } = {}) {
    this.baseUrl = `http://${host}:${port}`;
    this.retries = retries;
    this.retryDelay = retryDelay;
    this.timeout = timeout;
  }

  /**
   * Execute a raw RPC command.
   * @param {string} cmd - RPC command (e.g., "balance", "send address:MxG.. amount:1")
   * @returns {Promise<Object>} Full parsed JSON response
   * @throws {MinimaError} When status is false
   * @throws {MinimaConnectionError} When node is unreachable after retries
   */
  async command(cmd) {
    const encoded = encodeURIComponent(cmd);
    const url = `${this.baseUrl}/${encoded}`;

    let lastError;
    for (let attempt = 0; attempt < this.retries; attempt++) {
      try {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), this.timeout);

        const resp = await fetch(url, { signal: controller.signal });
        clearTimeout(timer);

        const text = await resp.text();
        const result = JSON.parse(text);

        if (!result.status) {
          throw new MinimaError(result.error || 'Unknown RPC error');
        }

        return result;
      } catch (e) {
        if (e instanceof MinimaError) throw e;
        lastError = e;
        if (attempt < this.retries - 1) {
          await sleep(this.retryDelay * (attempt + 1));
        }
      }
    }

    throw new MinimaConnectionError(
      `Failed to connect to ${this.baseUrl} after ${this.retries} attempts: ${lastError?.message}`
    );
  }

  /**
   * Get wallet balances with safe field naming.
   *
   * The dangerous 'total' field (token max supply) is moved to
   * supply.total so it cannot be confused with wallet balance.
   *
   * @param {string} [tokenid] - Optional token ID filter
   * @returns {Promise<Array<{
   *   token: string,
   *   tokenid: string,
   *   sendable: string,
   *   confirmed: string,
   *   unconfirmed: string,
   *   coins: string,
   *   supply: { total: string }
   * }>>}
   */
  async balance(tokenid) {
    const result = await this.command('balance');
    let entries = result.response || [];

    const normalized = entries.map(entry => ({
      token: entry.token || '',
      tokenid: entry.tokenid || '',
      sendable: entry.sendable || '0',
      confirmed: entry.confirmed || '0',
      unconfirmed: entry.unconfirmed || '0',
      coins: entry.coins || '0',
      supply: {
        total: entry.total || '0',
      },
    }));

    if (tokenid) {
      return normalized.filter(b => b.tokenid === tokenid);
    }
    return normalized;
  }

  /**
   * Get a simple balance summary for one token.
   * @param {string} [tokenid='0x00'] - Token ID
   * @returns {Promise<{ sendable: number, confirmed: number, unconfirmed: number, coins: number }>}
   */
  async balanceSummary(tokenid = '0x00') {
    const balances = await this.balance(tokenid);
    if (!balances.length) {
      return { sendable: 0, confirmed: 0, unconfirmed: 0, coins: 0 };
    }
    const b = balances[0];
    return {
      sendable: parseFloat(b.sendable) || 0,
      confirmed: parseFloat(b.confirmed) || 0,
      unconfirmed: parseFloat(b.unconfirmed) || 0,
      coins: parseInt(b.coins, 10) || 0,
    };
  }

  /**
   * Get node status with numeric fields parsed.
   * @returns {Promise<{
   *   version: string,
   *   chainHeight: number,
   *   block: number,
   *   devices: number,
   *   mempool: number,
   *   uptime: string,
   *   raw: Object
   * }>}
   */
  async status() {
    const result = await this.command('status');
    const resp = result.response || {};
    const chain = resp.chain || {};
    const txpow = resp.txpow || {};

    return {
      version: resp.version || '',
      chainHeight: safeInt(resp.length),
      block: safeInt(chain.block),
      devices: safeInt(resp.devices),
      mempool: safeInt(txpow.mempool),
      uptime: chain.date || '',
      raw: resp,
    };
  }

  /**
   * Send Minima or tokens to an address.
   * @param {string} address - Destination (MxG... or 0x...)
   * @param {number|string} amount - Amount to send
   * @param {Object} [options]
   * @param {string} [options.tokenid] - Token ID (default: native Minima)
   * @param {number} [options.split] - Split into N outputs (1-10)
   * @param {number|string} [options.burn] - Burn amount as fee
   * @returns {Promise<{ txpowid: string, explorerUrl: string, block: string, date: string, raw: Object }>}
   */
  async send(address, amount, { tokenid, split, burn } = {}) {
    let cmd = `send address:${address} amount:${amount}`;
    if (tokenid) cmd += ` tokenid:${tokenid}`;
    if (split) cmd += ` split:${split}`;
    if (burn) cmd += ` burn:${burn}`;

    const result = await this.command(cmd);
    const resp = result.response || {};
    const txpowid = resp.txpowid || '';

    return {
      txpowid,
      explorerUrl: `https://explorer.minima.global/transactions/${txpowid}`,
      block: resp.header?.block || '',
      date: resp.header?.date || '',
      raw: resp,
    };
  }

  /**
   * Hash data using Minima's Keccak-256.
   * @param {string} data - String or 0x-prefixed hex data
   * @returns {Promise<{ input: string, hash: string }>}
   */
  async hash(data) {
    const result = await this.command(`hash data:${data}`);
    const resp = result.response || {};
    return {
      input: resp.input || '',
      hash: resp.hash || '',
    };
  }

  /**
   * Generate 256-bit cryptographic random value.
   * @returns {Promise<string>} 0x-prefixed random hex
   */
  async random() {
    const result = await this.command('random');
    return result.response?.random || '';
  }

  /**
   * List all tokens known to this node.
   * @returns {Promise<Array<{ tokenid: string, name: string, supplyTotal: string, decimals: number, scale: number }>>}
   */
  async tokens() {
    const result = await this.command('tokens');
    const entries = result.response || [];

    return entries.map(entry => {
      const token = entry.token;
      const name = typeof token === 'string' ? token : (token?.name || String(token));

      return {
        tokenid: entry.tokenid || '',
        name,
        supplyTotal: entry.total || '0',
        decimals: safeInt(entry.decimals),
        scale: safeInt(entry.scale),
      };
    });
  }

  /**
   * Get the node's default receiving address.
   * @returns {Promise<{ address: string, miniaddress: string, publickey: string }>}
   */
  async getaddress() {
    const result = await this.command('getaddress');
    const resp = result.response || {};
    return {
      address: resp.address || '',
      miniaddress: resp.miniaddress || '',
      publickey: resp.publickey || '',
    };
  }

  /**
   * Get Maxima identity and contact details.
   * @returns {Promise<{ name: string, publickey: string, mls: string, p2pidentity: string, localidentity: string, contact: string }>}
   */
  async maximaInfo() {
    const result = await this.command('maxima action:info');
    const resp = result.response || {};
    return {
      name: resp.name || '',
      publickey: resp.publickey || '',
      mls: resp.mls || '',
      p2pidentity: resp.p2pidentity || '',
      localidentity: resp.localidentity || '',
      contact: resp.contact || '',
    };
  }

  /**
   * List Maxima contacts.
   * @returns {Promise<Array<{ id: number, name: string, publickey: string, address: string, lastseen: string, samechain: boolean }>>}
   */
  async contacts() {
    const result = await this.command('maxcontacts action:list');
    const entries = result.response || [];

    return entries.map(entry => ({
      id: entry.id || 0,
      name: entry.extradata?.name || '',
      publickey: entry.publickey || '',
      address: entry.currentaddress || '',
      lastseen: entry.date || '',
      samechain: entry.samechain || false,
    }));
  }
}

/** @param {string|number} val */
function safeInt(val) {
  const n = parseInt(val, 10);
  return isNaN(n) ? 0 : n;
}

/** @param {number} ms */
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export { MinimaClient, MinimaError, MinimaConnectionError };

File: integration/python/minima_client.py
"""
Minima RPC Client for Python

Safe, typed interface to a Minima node's RPC endpoint.
Normalizes responses so the default API path prevents common mistakes.

Usage:
    from minima_client import MinimaClient

    client = MinimaClient()           # default: localhost:9005
    bal = client.balance()            # normalized, no 'total' trap
    status = client.status()
    tx = client.send("MxG08...", 1)   # returns tx with explorer link
"""

import json
import time
import urllib.request
import urllib.parse
import urllib.error


class MinimaError(Exception):
    """Raised when an RPC command returns status=false."""
    pass


class MinimaConnectionError(Exception):
    """Raised when the node is unreachable after retries."""
    pass


class MinimaClient:
    """HTTP client for Minima RPC with retries and normalized responses."""

    def __init__(self, host="localhost", port=9005, retries=3, retry_delay=1.0):
        self.base_url = f"http://{host}:{port}"
        self.retries = retries
        self.retry_delay = retry_delay

    def command(self, cmd):
        """
        Execute a raw RPC command. Returns the parsed JSON response.

        Args:
            cmd: RPC command string (e.g., "balance", "send address:MxG.. amount:1")

        Returns:
            dict: Full parsed JSON response {"status": bool, "response": ...}

        Raises:
            MinimaError: When status is false
            MinimaConnectionError: When node is unreachable after retries
        """
        encoded = urllib.parse.quote(cmd, safe="")
        url = f"{self.base_url}/{encoded}"

        last_error = None
        for attempt in range(self.retries):
            try:
                req = urllib.request.Request(url)
                with urllib.request.urlopen(req, timeout=30) as resp:
                    raw = resp.read().decode("utf-8")
                    result = json.loads(raw)

                    if not result.get("status", False):
                        raise MinimaError(result.get("error", "Unknown RPC error"))

                    return result

            except (urllib.error.URLError, ConnectionError, OSError) as e:
                last_error = e
                if attempt < self.retries - 1:
                    time.sleep(self.retry_delay * (attempt + 1))

        raise MinimaConnectionError(
            f"Failed to connect to {self.base_url} after {self.retries} attempts: {last_error}"
        )

    def balance(self, tokenid=None):
        """
        Get wallet balances with safe field naming.

        The dangerous 'total' field (token max supply) is moved to
        supply.total so it cannot be confused with wallet balance.

        Args:
            tokenid: Optional token ID filter (e.g., "0x00" for Minima)

        Returns:
            list[dict]: Normalized balances, each with:
                - token (str): Token display name
                - tokenid (str): Token identifier
                - sendable (str): Available to spend — PRIMARY BALANCE
                - confirmed (str): Full wallet balance (includes locked)
                - unconfirmed (str): Pending incoming
                - coins (str): Number of UTXOs
                - supply (dict): {"total": str} — token max supply, NOT balance
        """
        result = self.command("balance")
        entries = result.get("response", [])

        normalized = []
        for entry in entries:
            item = {
                "token": entry.get("token", ""),
                "tokenid": entry.get("tokenid", ""),
                "sendable": entry.get("sendable", "0"),
                "confirmed": entry.get("confirmed", "0"),
                "unconfirmed": entry.get("unconfirmed", "0"),
                "coins": entry.get("coins", "0"),
                "supply": {
                    "total": entry.get("total", "0"),
                },
            }
            normalized.append(item)

        if tokenid:
            normalized = [b for b in normalized if b["tokenid"] == tokenid]

        return normalized

    def balance_summary(self, tokenid="0x00"):
        """
        Get a simple balance summary for one token.

        Returns:
            dict: {sendable, confirmed, unconfirmed, coins} as floats
        """
        balances = self.balance(tokenid=tokenid)
        if not balances:
            return {"sendable": 0.0, "confirmed": 0.0, "unconfirmed": 0.0, "coins": 0}

        b = balances[0]
        return {
            "sendable": float(b["sendable"]),
            "confirmed": float(b["confirmed"]),
            "unconfirmed": float(b["unconfirmed"]),
            "coins": int(b["coins"]),
        }

    def status(self):
        """
        Get node status with numeric fields parsed.

        Returns:
            dict: Normalized status with parsed integers where appropriate:
                - version (str)
                - chain_height (int): Parsed from response.length
                - block (int): Current block number
                - devices (int)
                - mempool (int): Pending transactions
                - uptime (str): Human-readable date
                - raw (dict): Full unmodified response
        """
        result = self.command("status")
        resp = result.get("response", {})
        chain = resp.get("chain", {})
        txpow = resp.get("txpow", {})

        return {
            "version": resp.get("version", ""),
            "chain_height": _safe_int(resp.get("length", "0")),
            "block": _safe_int(chain.get("block", "0")),
            "devices": _safe_int(resp.get("devices", "0")),
            "mempool": _safe_int(txpow.get("mempool", "0")),
            "uptime": chain.get("date", ""),
            "raw": resp,
        }

    def send(self, address, amount, tokenid=None, split=None, burn=None):
        """
        Send Minima or tokens to an address.

        Args:
            address: Destination (MxG... or 0x...)
            amount: Amount to send
            tokenid: Token ID (default: 0x00 for native Minima)
            split: Split into N output coins (1-10)
            burn: Burn amount as fee

        Returns:
            dict: {txpowid, explorer_url, block, date, raw}
        """
        cmd = f"send address:{address} amount:{amount}"
        if tokenid:
            cmd += f" tokenid:{tokenid}"
        if split:
            cmd += f" split:{split}"
        if burn:
            cmd += f" burn:{burn}"

        result = self.command(cmd)
        resp = result.get("response", {})
        txpowid = resp.get("txpowid", "")

        return {
            "txpowid": txpowid,
            "explorer_url": f"https://explorer.minima.global/transactions/{txpowid}",
            "block": resp.get("header", {}).get("block", ""),
            "date": resp.get("header", {}).get("date", ""),
            "raw": resp,
        }

    def hash(self, data):
        """
        Hash data using Minima's Keccak-256.

        Args:
            data: String or 0x-prefixed hex data

        Returns:
            dict: {input, hash}
        """
        result = self.command(f"hash data:{data}")
        resp = result.get("response", {})
        return {
            "input": resp.get("input", ""),
            "hash": resp.get("hash", ""),
        }

    def random(self):
        """
        Generate 256-bit cryptographic random value.

        Returns:
            str: 0x-prefixed random hex string
        """
        result = self.command("random")
        return result.get("response", {}).get("random", "")

    def tokens(self):
        """
        List all tokens known to this node.

        Returns:
            list[dict]: Each with {tokenid, name, supply_total, decimals, scale}
                        Note: supply_total is token max supply, NOT wallet balance.
        """
        result = self.command("tokens")
        entries = result.get("response", [])

        normalized = []
        for entry in entries:
            token = entry.get("token", "")
            name = token if isinstance(token, str) else token.get("name", str(token))

            normalized.append({
                "tokenid": entry.get("tokenid", ""),
                "name": name,
                "supply_total": entry.get("total", "0"),
                "decimals": _safe_int(entry.get("decimals", "0")),
                "scale": _safe_int(entry.get("scale", "0")),
            })

        return normalized

    def getaddress(self):
        """
        Get the node's default receiving address.

        Returns:
            dict: {address, miniaddress, publickey}
        """
        result = self.command("getaddress")
        resp = result.get("response", {})
        return {
            "address": resp.get("address", ""),
            "miniaddress": resp.get("miniaddress", ""),
            "publickey": resp.get("publickey", ""),
        }

    def maxima_info(self):
        """
        Get Maxima identity and contact details.

        Returns:
            dict: {name, publickey, mls, p2pidentity, localidentity, contact}
        """
        result = self.command("maxima action:info")
        resp = result.get("response", {})
        return {
            "name": resp.get("name", ""),
            "publickey": resp.get("publickey", ""),
            "mls": resp.get("mls", ""),
            "p2pidentity": resp.get("p2pidentity", ""),
            "localidentity": resp.get("localidentity", ""),
            "contact": resp.get("contact", ""),
        }

    def contacts(self):
        """
        List Maxima contacts.

        Returns:
            list[dict]: Each with {id, name, publickey, address, lastseen, samechain}
        """
        result = self.command("maxcontacts action:list")
        entries = result.get("response", [])

        normalized = []
        for entry in entries:
            extra = entry.get("extradata", {})
            normalized.append({
                "id": entry.get("id", 0),
                "name": extra.get("name", ""),
                "publickey": entry.get("publickey", ""),
                "address": entry.get("currentaddress", ""),
                "lastseen": entry.get("date", ""),
                "samechain": entry.get("samechain", False),
            })

        return normalized


def _safe_int(val):
    """Safely parse a string to int, returning 0 on failure."""
    try:
        return int(val)
    except (ValueError, TypeError):
        return 0
